datafusion/optimizer/Cargo.toml

```diff
@@ -44,6 +44,7 @@ async-trait = { workspace = true }
 chrono = { workspace = true }
 datafusion-common = { workspace = true }
 datafusion-expr = { workspace = true }
+datafusion-functions-array = { workspace = true }
```

Comment:
I am not sure about this new dependency -- it means that using the optimizer will require bringing in the physical exprs, etc and users (like dask-sql) that only want the planner code would bring in substantial unused code. 

I realize the reason this is required is to preserve the semantics of the existing rewrite pass in `datafusion/optimizer/src/analyzer/rewrite_expr.rs`. I wonder if we can somehow avoid adding this new dependency 🤔

Reply:
Maybe we can split udf to logical expr and physcial expr, and export only logical expr conditionally (with `#[cfg(feature = "array_expressions")]`) to optimizer, so we can do optimizing for udfs, but only import the necessary one for user.

Reply:
Is it possible to avoid importing `function-arrays` if we get udf via `self.context_provider.get_function_meta(&name)` ?

Reply:
> Is it possible to avoid importing `function-arrays` if we get udf via `self.context_provider.get_function_meta(&name)` ?

wow, This may be feasible

Reply:
hi @jayzhan211 , Do you think my current approach to the analyzer is feasible? I move array analyzer to `datafusion-functions-array` crate, add array analyzer to `analyzer rules` if `array_expression` featrue is enable

Reply:
Probably not a good idea, it seems like duplicated code without any good reason.